Use pkg/webhook along with build webhook implementation

shashwathi opened the pull request:
Address #378 

## Proposed Changes
  * Do not use custom build webhook implementation
  * Implement validate and default methods for all build CRDs and use pkg/wehbook controller.

**FYI**
 I have not added any tests in this PR. If this change does not get 👎 then I will add tests to this PR and also delete custom webhook code.

Goal of this PR is to get conversation started about this change

Lot of validation in build CRDs included calling k8s API to validate object exist like
secrets, build templates. I have not included that change here because I noticed that Serving did not have any similar validations. 
My question is does build CRD want that validation still? 

Thoughts or concerns @ImJasonH @mattmoor @pivotal-nader-ziada ?

ImJasonH commented on pkg/apis/build/v1alpha1/build_defaults.go:
```diff
@@ -0,0 +1,7 @@
+package v1alpha1
+
+// SetDefaults for build
+func (b *Build) SetDefaults() {
```
Can you add `// TODO(shashwathi): Set default timeout to 10s` here?

ImJasonH commented on pkg/apis/build/v1alpha1/build_template_validation.go:
```diff
@@ -0,0 +1,125 @@
+		return err
+	}
+	return nil
+	// return validatePlaceholders(tmpl.TemplateSpec().Steps)
```
I think this was erroneously commented out.

ImJasonH commented on pkg/apis/build/v1alpha1/build_validation.go:
```diff
@@ -0,0 +1,182 @@
+	}
+	if bs.Template != nil && len(bs.Steps) > 0 {
+		return apis.ErrMissingField("b.spec.template").Also(apis.ErrMissingField("b.spec.steps"))
+		//return validationError("TemplateAndSteps", "build cannot specify both template and steps")
```
These `return validationError` lines can just be deleted.

ImJasonH commented on pkg/apis/build/v1alpha1/build_validation.go:
```diff
@@ -0,0 +1,182 @@
+		return err
+	}
+
+	// Need kubeclient to fetch template
```
Yeah I think the current thinking is that we'll only do simple validation in the webhook, and leave it up to the controller to do full validation with resolved objects like the template, service account, secrets, etc.

I'm not sure there's a good story yet for surfacing validation errors outside of webhook checks though, we'd probably need to have some new condition type to express the `Invalid` state.

ImJasonH commented on pkg/apis/build/v1alpha1/build_validation.go:
```diff
@@ -0,0 +1,182 @@
+
+// ValidateSecrets for build
+func (bs *BuildSpec) ValidateSecrets() *apis.FieldError {
+	// TODO: move this logic into set defaults
```
+1 good call

shashwathi commented:
@ImJasonH 

I like the idea of 2 controllers for now. Updated code to split the validation between the 2 webhooks. Added tests and cleaned up bunch of common dependencies.

shashwathi commented:
/assign @ImJasonH 
/assign @pivotal-nader-ziada

mattmoor commented on cmd/webhook/main.go:
```diff
@@ -78,12 +83,27 @@ func main() {
+		Options: pkgoptions,
+		Handlers: map[schema.GroupVersionKind]pkgwebhook.GenericCRD{
+			v1alpha1.SchemeGroupVersion.WithKind("Build"):                &v1alpha1.Build{},
+			v1alpha1.SchemeGroupVersion.WithKind("ClusterBuildTemplate"): &v1alpha1.ClusterBuildTemplate{},
```
Can you confirm that the common infrastructure works properly with cluster-scoped resources, since AFAIK this is the first one we're trying :)

shashwathi replied:
Cool 👍 
I have added e2e test for build with template and build with cluster template as well. Going forward we will have better coverage for these features

mattmoor commented on pkg/apis/build/v1alpha1/build_defaults.go:
```diff
@@ -0,0 +1,20 @@
+)
+
+// DefaultTime is 10min
+const DefaultTime = 10 * time.Minute
```
nit: DefaultTimeout

ImJasonH commented on pkg/apis/build/v1alpha1/build_defaults_test.go:
```diff
@@ -0,0 +1,16 @@
+	"testing"
+)
+
+func TestSetDefault(t *testing.T) {
```
Can you also test that a build that specifies both values doesn't have its values overwritten?

ImJasonH commented on pkg/apis/build/v1alpha1/build_template_defaults.go:
```diff
@@ -0,0 +1,6 @@
+package v1alpha1
+
+// SetDefaults for build template
+func (b *BuildTemplate) SetDefaults() {
```
Can we just put this into `build_template_types.go` ? This separate file doesn't add much.

ImJasonH commented on pkg/apis/build/v1alpha1/build_template_validation_test.go:
```diff
@@ -0,0 +1,127 @@
+		reason: "long name",
+		tmpl: &BuildTemplate{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "*********************************************************************************************",
```
`strings.Repeat("x", maxLength+1)` is self-documenting and self-updating if we update maxLength.

shashwathi replied:
I missed my TODO there. Good catch

ImJasonH commented on pkg/apis/build/v1alpha1/build_template_validation.go:
```diff
@@ -0,0 +1,91 @@
+		}
+	}
+
+	if len(name) > 63 {
```
Can we extract this to a const?

ImJasonH commented on pkg/apis/build/v1alpha1/build_template_validation.go:
```diff
@@ -0,0 +1,91 @@
+	seen := map[string]struct{}{}
+	for _, p := range params {
+		if _, ok := seen[p.Name]; ok {
+			return apis.ErrMultipleOneOf("ParamName")
```
I don't think `MultipleOneOf` is the correct error, I think that error is intended to communicate "you can specify X or Y, but not both"

https://godoc.org/github.com/knative/pkg/apis#ErrMultipleOneOf

(`ErrMissingOneOf` is "you must specify X or Y, but not neither")

Maybe `ErrInvalidKeyName`? https://godoc.org/github.com/knative/pkg/apis#ErrInvalidKeyName

ImJasonH commented on pkg/apis/build/v1alpha1/clutser_build_template_validation_test.go:
```diff
@@ -0,0 +1,111 @@
+package v1alpha1
```
typo in filename: `cluster_build_template_validation_test.go`

ImJasonH commented on pkg/apis/build/v1alpha1/clutser_build_template_validation_test.go:
```diff
@@ -0,0 +1,111 @@
+		reason string // if "", expect success.
+	}{{
+		desc: "Single named step",
+		tmpl: &ClusterBuildTemplate{
```
I think we can condense both of these test tables into one, and avoid duplication:

```
func TestValidateBuildTemplate(t *testing.T) {
  for _, c := range []struct {
    desc string
    spec BuildTemplateSpec
    reason string
  } {{
   ...
  }} {
    t.Run("namespaced-"+c.desc, func(t *testing.T) {
      tmpl := &BuildTemplate{Spec: c.spec},
      tmpl.Validate()
      ...
    })
    t.Run("cluster-namespaced-"+c.desc, func(t *testing.T) {
      ctmpl := &ClusterBuildTemplate{Spec: c.spec},
      ctmpl.Validate()
      ...
    })
  }
}
```
